src/models/kinflated_gamma_poisson_model.py

```diff
@@ -0,0 +1,602 @@
+
+        Args:
+          h: Actual frequency histogram.  h[i] is the number of people
+             reached exactly i+1 times, except for h[-1] which is the
+             number of people reached len(h) or more times.
+          I: Total number of impressions purchased.
+          N: Total number of unique users.
+          dist:   A KInflatedGammaPoisson distribution.
+        Returns:
+          chi^2 distance between the actual and expected histograms, plus
+          a term for weighting the difference in 1+ reach.
+        """
+        # Estimate total number of potential impressions
+        Imax = N * dist.expected_value()
+        if Imax <= I:
+            return np.sum(np.array(h) ** 2)
+
+        freqs = N * dist.kreach(np.arange(1, len(h)), I / Imax)
+        kplus = N * dist.kplusreach(len(h), I / Imax)
+        hbar = np.array(list(freqs) + [kplus])
+
+        obj = np.sum((hbar - h) ** 2 / (hbar + 1e-6))
+
+        if np.isnan(obj):
+            logging.vlog(2, f"alpha {alpha} beta {beta} N {N} Imax {Imax}")
+            logging.vlog(2, f"h    {h}")
+            logging.vlog(2, f"hbar {hbar}")
+            raise RuntimeError("Invalid value of objective function")
+
+        return obj
+
+    def _exponential_poisson_reach(self, I: float, N: float, beta: float) -> float:
+        """Estimates reach in an exponential poisson model.
+
+        Estimates reach when I impressions are purchased from a population
+        of N users where the impression count of each user is distributed
+        according to the exponential poisson distribution with parameter beta.
+
+        Under the exponential poisson model, the probability that a
+        user will have n impressions is (1 - p) * p^{n-1}, where p =
+        beta/(beta + 1).  This can be worked out by using the
+        equivalence between the gamma poisson and the negative
+        binomial and setting alpha = r = 1.
+
+        The probability that a user is not reached, given that a fraction q
+        of available impressions is purchased, is given by
+           sum_{n=1}^{infinity} (1 - q)^n f(n),
+        where f(n) is the PMF of the above exponential poisson distribution.
+        This simplifies to (1 - p)(1 - q) / (1 - p(1 - q)).
+
+        The probability that a user is reached at least once is one minus
+        the above quantity.  Finally, the value of q is taken to be
+        I / ((beta + 1) * N), as the average number of impressions that a
+        user will have is beta + 1.
+
+        Args:
+          I:  The hypothetical number of impressions purchased.
+          N:  The audience size.
+          beta: The parameter of the exponential-poisson distribution.
+        Returns:
+          The expected reach.
+        """
+        return N * I * (beta + 1) / (N + beta * (N + I))
+
+    def _exponential_poisson_beta(self, I: float, N: float, R: float) -> float:
+        """Estimate beta given impressions, audience size and reach."""
```

Comment:
Perhaps clarify that estimate beta by method of first moment, i.e., backsolving self._exponential_poisson_reach(I, N, beta) = R

Reply:
Done.